File: AlgorithmLcr/Try2-Known component types, quantities and parameters/netgraph_id/adjacency.py
"""Unified adjacency-matrix output (../../OUTPUT_FORMAT.md).

Expands a candidate ``Network`` (structure + component assignment) into the
canonical upper-triangle adjacency matrix: ``rows[i][j]`` (i < j) is the
vector of all edges directly connecting nodes i and j.  Nodes 0 and 1 are
the one-port terminals.

Edge follows the root spec: (type, parameter, dcr).  Try2 components carry
their known values; inductors contribute their series DCR, R/C edges carry
dcr = 0.  The expansion is lossless: ``Structure.mult`` is already the
flattened upper triangle (slot order = row-major), and edge instances are
contiguous per slot (``Structure.slot_of_instances``).
"""

from __future__ import annotations

from dataclasses import dataclass

from .components import ComponentSet
from .graph import Network, slot_list
from .selector import Candidate


@dataclass(frozen=True)
class Edge:
    """One 2-terminal element; see OUTPUT_FORMAT.md section 1."""

    type: str          # "R" | "L" | "C"
    parameter: float   # R[ohm] / L[H] / C[F]
    dcr: float = 0.0   # series DC resistance of L; 0 unless type == "L"


def _fmt_edge(e: Edge) -> str:
    if e.type == "L" and e.dcr != 0.0:
        return f"L {e.parameter:.3e} dcr {e.dcr:.3e}"
    return f"{e.type} {e.parameter:.3e}"


class Adjacency:
    """Strict upper-triangle adjacency matrix of vector<Edge> (spec sec.2)."""

    def __init__(self, V: int):
        if V < 2:
            raise ValueError("a one-port needs at least the 2 terminal nodes")
        self.V = V
        self.rows = [[[] for _ in range(V - 1 - i)] for i in range(V)]

    def slot(self, i: int, j: int) -> list[Edge]:
        """All edges directly connecting i and j; requires i < j."""
        if not (0 <= i < j < self.V):
            raise ValueError(f"slot ({i},{j}) outside upper triangle of V={self.V}")
        return self.rows[i][j - i - 1]

    def add(self, i: int, j: int, edge: Edge) -> None:
        """Append an undirected edge; {i, j} order is normalized."""
        if i == j:
            raise ValueError("self loops are not part of the format")
        if i > j:
            i, j = j, i
        self.slot(i, j).append(edge)

    @property
    def n_edges(self) -> int:
        return sum(len(cell) for row in self.rows for cell in row)

    def occupied(self) -> list[tuple[int, int, list[Edge]]]:
        """Non-empty slots in row-major upper-triangle order (spec sec.4)."""
        out = []
        for i in range(self.V):
            for k, cell in enumerate(self.rows[i]):
                if cell:
                    out.append((i, i + k + 1, cell))
        return out

    def format_block(self, label: int | str | None = None,
                     extra_lines: list[str] | None = None) -> str:
        """Unified print form (spec sec.4); label = candidate rank when given."""
        head = f"adjacency[{label}] " if label is not None else "adjacency "
        lines = [f"{head}V={self.V} (ports 0,1):"]
        for i, j, edges in self.occupied():
            lines.append(f"  ({i},{j}): " + " | ".join(_fmt_edge(e) for e in edges))
        if extra_lines:
            lines.extend("  " + s for s in extra_lines)
        return "\n".join(lines)


def network_to_adjacency(network: Network, compset: ComponentSet) -> Adjacency:
    """Expand a candidate network into the unified adjacency matrix.

    Slot order (row-major upper triangle) and in-slot edge order (instance
    order) are both deterministic; the mapping is one-to-one.
    """
    V = network.structure.V
    slots = slot_list(V)
    soi = network.structure.slot_of_instances()
    comps = compset.components
    adj = Adjacency(V)
    for t, comp_idx in enumerate(network.assign):
        i, j = slots[soi[t]]
        comp = comps[comp_idx]
        adj.add(i, j, Edge(comp.kind, comp.value, comp.dcr))
    return adj


def candidate_to_adjacency(cand: Candidate, compset: ComponentSet) -> Adjacency:
    """Adjacency matrix of a ranked candidate (spec sec.4: per representative)."""
    return network_to_adjacency(cand.network, compset)

File: AlgorithmLcr/Try2-Known component types, quantities and parameters/cppversion/src/adjacency.cpp
#include "adjacency.hpp"

#include <cstdio>
#include <stdexcept>

namespace ng {

namespace {

std::string fmtEdge(const Edge& e) {
    char buf[96];
    if (e.type == 'L' && e.dcr != 0.0) {
        std::snprintf(buf, sizeof(buf), "L %.3e dcr %.3e", e.parameter, e.dcr);
    } else {
        std::snprintf(buf, sizeof(buf), "%c %.3e", e.type, e.parameter);
    }
    return buf;
}

}  // namespace

Adjacency::Adjacency(int V) : V_(V) {
    if (V < 2) throw std::invalid_argument("a one-port needs at least the 2 terminal nodes");
    rows_.resize((size_t)V);
    for (int i = 0; i < V; ++i) rows_[i].resize((size_t)V - 1 - i);
}

std::vector<Edge>& Adjacency::slot(int i, int j) {
    if (!(0 <= i && i < j && j < V_))
        throw std::invalid_argument("slot outside upper triangle");
    return rows_[i][(size_t)(j - i - 1)];
}

const std::vector<Edge>& Adjacency::slot(int i, int j) const {
    return const_cast<Adjacency*>(this)->slot(i, j);
}

void Adjacency::add(int i, int j, Edge edge) {
    if (i == j) throw std::invalid_argument("self loops are not part of the format");
    if (i > j) std::swap(i, j);
    slot(i, j).push_back(edge);
}

int Adjacency::nEdges() const {
    int n = 0;
    for (const auto& row : rows_)
        for (const auto& cell : row) n += (int)cell.size();
    return n;
}

std::vector<std::tuple<int, int, const std::vector<Edge>*>> Adjacency::occupied() const {
    std::vector<std::tuple<int, int, const std::vector<Edge>*>> out;
    for (int i = 0; i < V_; ++i)
        for (int k = 0; k < V_ - 1 - i; ++k)
            if (!rows_[i][(size_t)k].empty())
                out.push_back({i, i + k + 1, &rows_[i][(size_t)k]});
    return out;
}

std::string Adjacency::formatBlock(const std::string& label,
                                   const std::vector<std::string>& extraLines) const {
    std::string out;
    out += label.empty() ? "adjacency " : ("adjacency[" + label + "] ");
    out += "V=" + std::to_string(V_) + " (ports 0,1):\n";
    for (const auto& [i, j, edges] : occupied()) {
        char head[32];
        std::snprintf(head, sizeof(head), "  (%d,%d): ", i, j);
        out += head;
        for (size_t q = 0; q < edges->size(); ++q) {
            if (q) out += " | ";
            out += fmtEdge((*edges)[q]);
        }
        out += "\n";
    }
    for (const auto& s : extraLines) out += "  " + s + "\n";
    if (!out.empty()) out.pop_back();  // trailing newline
    return out;
}

Adjacency networkToAdjacency(const Network& network, const ComponentSet& compset) {
    return networkToAdjacency(network, compset.components());
}

Adjacency networkToAdjacency(const Network& network, const std::vector<Component>& comps) {
    const int V = network.structure.V;
    const auto& slots = slotList(V);
    const auto soi = network.structure.slotOfInstances();
    Adjacency adj(V);
    for (size_t t = 0; t < network.assign.size(); ++t) {
        auto [i, j] = slots[soi[t]];
        const Component& c = comps[network.assign[t]];
        adj.add(i, j, Edge{c.kind, c.value, c.dcr});
    }
    return adj;
}

Adjacency candidateToAdjacency(const Candidate& cand, const ComponentSet& compset) {
    return networkToAdjacency(cand.network, compset);
}

Adjacency candidateToAdjacency(const Candidate& cand, const std::vector<Component>& comps) {
    return networkToAdjacency(cand.network, comps);
}

}  // namespace ng
